### PR TITLE
Implement Quantum Storage Controller Network

### PR DESCRIPTION
**Pull request by ghzdude**
## What
Implements the Quantum Storage Controller, a block which combines the inventories of connected Super/Quantum Tanks/Chests to allow I/O from attached storages. This PR continues the work of the `ds-quantum-storage-improved` branch.

## Implementation Details
- Quantum Storage MTEs use IDs 1757 - 1760
- Implements interfaces `IQuantumStorage` and `IQuantumController`.
- Implements abstract class  `MetaTileEntityQuantumStorage` with logic to handle dis/connecting from the Quantum Controller.
- Quantum Controller has logic to rebuild it's network and a handler to handle I/O from attached storages
- Item I/O from the Quantum Controller handler is working, but could be improved I think.
- Rendering of the Quantum Controller should be discussed, if current rendering is not desired.
- Creative Quantum Tank/Chest works with the storage controller as well
- Creative Tank/Chest will only output from the storage controller if it is active
- Normal Creative Tank/Chest behavior (auto-output to the output face) is disabled if connected to a storage controller

## To-do

- [x] Implement MTEs
  - [x] Quantum Storage Controller
  - [x] Quantum Storage Extender
  - [x] Quantum Storage Proxy
- [x] Add textures for...
  - [x] Quantum Storage Controller
  - [x] Quantum Storage Extender
  - [x] Quantum Storage Proxy
- [x] add recipes for...
  - [x] Quantum Storage Controller
  - [x] Quantum Storage Extender
  - [x] Quantum Storage Proxy
- [x] change the connection icon to the GT logo and put it in the bottom right corner
- [ ] add energy usage
  - [x] TOP
  - [ ] Hwyla
- [x] add indicator overlay
- [x] add controller highlight rendering to connected qchests/qtanks

## Outcome
Players can connect several super/quantum chests/tanks together and I/O from them all.

## Additional Information
https://github.com/GregTechCEu/GregTech/assets/44148655/e2aa6400-d1d2-4284-860e-58e24dfbe9dc

## Potential Compatibility Issues
Existing instances of super/quantum tanks/chests should not be affected,  otherwise no other compatibility issues afaik.

### CODE REVIEW

**Comment by IntegerLimit** [comment]
Completely rendering fluid drawers (and by extension, storage drawers) useless. Yes!

**Comment by ghzdude** [comment]
this pr will be ready for review once the 2.8 branch is merged into master and I rebase it if need be

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    public void onPlacement() {
+        // add to the network if an adjacent block is part of a network
+        // use whatever we find first, merging networks is not supported
+        if (!getWorld().isRemote) {
+            tryFindNewNetwork();
+        }
+    }
+
+    private boolean tryFindNewNetwork() {
+        for (EnumFacing facing : EnumFacing.VALUES) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:115]: This could probably also be checked in `onNeighborChanged`, so that the case that I presented in discord could be fixed a bit better.

Or an option to use the Neighbor cache here.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            controllerPos = buf.readBlockPos();
+        }
+    }
+
+    @Override
+    public void receiveCustomData(int dataId, PacketBuffer buf) {
+        super.receiveCustomData(dataId, buf);
+        if (dataId == UPDATE_CONTROLLER_POS) {
+            this.controllerPos = buf.readBlockPos();
+        } else if (dataId == REMOVE_CONTROLLER) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:162]: This should also kill the `controller` right?

**Review by ALongStringOfNumbers — COMMENTED** [review]
The controller the tank/chest is connected to in the GUI is a nice touch, but if possible, it might be even nicer to have it as a clickable button that highlights the controller in world.

Also, I am having thoughts about the Controller having a GUI that lists all the fluids/items that are stored via connected tanks/chests.

Also, thoughts on the controller having a power upkeep requirement?

In addition, I think that the parts should be gated differently, but that is just me, we will have to see what everyone else thinks

Also, I am not sure about the need for the ULV super tanks/chests. The super chest 1 is available extremely early (basically only needs 4 LV circuits), so I think relying on our crates is fine until this point.
If it is decided to keep these, then they need recipes.

I encounter a slight error in linking, detailed more in discord. 

Could this PR make use of the Neighbor cache with MTE.getNeighbor, instead of calling GTUtility.getMTE

**Comment by ghzdude** [comment]
> The controller the tank/chest is connected to in the GUI is a nice touch, but if possible, it might be even nicer to have it as a clickable button that highlights the controller in world.
> 
:servernotforrendering: :trollface: (Honestly not a bad idea, but i'd have to figure out rendering to do that)
> Also, I am having thoughts about the Controller having a GUI that lists all the fluids/items that are stored via connected tanks/chests.
> 
I don't know about this, you could just as easily slap an AE2 storage bus on the controller or proxy to see all the items or fluids. 
I feel like trying to implement this will result in ~duping~ crafting station 2.0, and this could maybe wait for new MUI to be implemented.
> Also, thoughts on the controller having a power upkeep requirement?
> 
Hmm, this could be interesting. I'd probably add an "energy acceptor" type block to power the quantum controller network (to keep the sides of the controller or proxy open for I/O)
> Also, I am not sure about the need for the ULV super tanks/chests. The super chest 1 is available extremely early (basically only needs 4 LV circuits), so I think relying on our crates is fine until this point. If it is decided to keep these, then they need recipes.
> 
The idea behind the ULV chests/tanks is that they can compete with drawers from Storage Drawers, which mostly need wood to craft. For some ideas, I think the ULV chests/tanks could take some wrought iron, a bronze crate, and ULV-tier circuits,

**Comment by ghzdude** [comment]
new TODOs:
- [ ] (re)add texture for inactive quantum controller
- [ ] add energy usage to TOP/Waila
- [ ] textures and recipe for quantum energy acceptor
- [ ] change the connection icon to the GT logo and put it in the bottom right corner
- [ ] add controller highlight rendering to connected qchests/qtanks

Are we sure the ULV chests/tanks are unnecessary? I know that the LV super chest is cheap, so the ULV super chest doesn't have that much of a reason to exist. On the other hand, the LV super tank is a little more expensive requiring Polyethylene, so there's some room for the ULV tank.

To be honest, now that I think of it the ULV chests/tanks really aren't that necessary, since you can't make the quantum controller that early. Which means you could just use drums are other storage solutions in the meantime.

> In addition, I think that the parts should be gated differently, but that is just me, we will have to see what everyone else thinks

Can you elaborate on this a little more? Anyone can join the discussion as well.

**Comment by ghzdude** [comment]
> Also, I am not sure about the need for the ULV super tanks/chests. The super chest 1 is available extremely early (basically only needs 4 LV circuits), so I think relying on our crates is fine until this point.
> If it is decided to keep these, then they need recipes.

the ULV super chest/tank are removed now, i now deem them not necessary

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            this.controllerPos = BlockPos.fromLong(data.getLong("ControllerPos"));
+        }
+    }
+
+    protected ClickButtonWidget createConnectedGui(int y) {
+        connectedIcon = new ClickButtonWidget(151, y, 18, 18, "",
+                clickData -> {
+                    writeCustomData(GregtechDataCodes.LOCATE_CONTROLLER, buffer -> {
+                        buffer.writeBoolean(this.isConnected());
+                    });
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:239]: Instead of writing if the quantum part is connected in the custom data, and then always writing the custom data, why not just check if the quantum part is connected first, and then only write the packet if it is?

File: src/main/java/gregtech/common/metatileentities/MetaTileEntities.java [hunk 4]
```diff
@@ -222,6 +225,10 @@ public class MetaTileEntities {
     public static final MetaTileEntityRotorHolder[] ROTOR_HOLDER = new MetaTileEntityRotorHolder[6]; // HV, EV, IV, LuV, ZPM, UV
     public static final MetaTileEntityMufflerHatch[] MUFFLER_HATCH = new MetaTileEntityMufflerHatch[GTValues.UV + 1]; // LV-UV
     public static final MetaTileEntityFusionReactor[] FUSION_REACTOR = new MetaTileEntityFusionReactor[3];
+    public static MetaTileEntityQuantumStorageController QUANTUM_STORAGE_CONTROLLER;
+//    public static MetaTileEntityQuantumEnergyAcceptor QUANTUM_ENERGY_ACCEPTOR;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/MetaTileEntities.java:229]: Probably just remove this if it is not going to be implemented.

File: src/main/java/gregtech/common/metatileentities/MetaTileEntities.java [hunk 5]
```diff
@@ -977,12 +984,22 @@ public static void init() {
...
+        QUANTUM_STORAGE_CONTROLLER = registerMetaTileEntity(1757,
+                new MetaTileEntityQuantumStorageController(gregtechId("quantum_storage_controller")));
+        QUANTUM_STORAGE_PROXY = registerMetaTileEntity(1758,
+                new MetaTileEntityQuantumProxy(gregtechId("quantum_storage_proxy")));
+        QUANTUM_STORAGE_EXTENDER = registerMetaTileEntity(1759,
+                new MetaTileEntityQuantumExtender(gregtechId("quantum_storage_extender")));
+        // QUANTUM_ENERGY_ACCEPTOR = registerMetaTileEntity(1760,
+        // new MetaTileEntityQuantumEnergyAcceptor(gregtechId("quantum_energy_acceptor")));
+
+        // Super / Quantum Chests, IDs 1559-1574
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/MetaTileEntities.java:997]: Why was this changed? It seems like not a good idea to change the assigned ids of existing mtes.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+        }
+    }
+
+    @Override
+    public Pair<TextureAtlasSprite, Integer> getParticleTexture() {
+        return Pair.of(Textures.QUANTUM_CONTROLLER_ACTIVE.getParticleSprite(), getPaintingColorForRendering());
+    }
+
+    @Override
+    protected boolean openGUIOnRightClick() {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:140]: UI could be cool, listing out all the connected inventories and their contents and positions. But it doesn't have to be implemented for this PR, since the network should work as is.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+        // need to make sure it still exists
+        MetaTileEntity mte = GTUtility.getMetaTileEntity(getWorld(), pos);
+        if (mte instanceof IQuantumStorage<?>storage) {
+            storageInstances.put(pos, new WeakReference<>(storage));
+            return storage;
+        } else if (rebuild) {
+            // need to remove and rebuild
+            storagePositions.remove(pos);
+            rebuildNetwork();
+            return null;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:168]: Could rebuilding the network end up returning a valid storage in a new network? Do we have to take this into account?

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+
+    @Override
+    public void onRemoval() {
+        if (getWorld().isRemote) return;
+        isDead = true;
+        for (BlockPos pos : storagePositions) {
+            IQuantumStorage<?> storage = getStorage(pos, false);
+            if (storage != null) storage.setDisconnected();
+        }
+        handler.invalidate();
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:193]: Should we also clear the list of storage Positions?

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+        return isXValid && isYValid && isZValid;
+    }
+
+    @Override
+    public void onRemoval() {
+        if (getWorld().isRemote) return;
+        isDead = true;
+        for (BlockPos pos : storagePositions) {
+            IQuantumStorage<?> storage = getStorage(pos, false);
+            if (storage != null) storage.setDisconnected();
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:191]: This call could get a bit expensive, given that it is a loop over all the valid storages in the network, and each call to disconnect the storage sees the storage attempt to find a new network. Do you have any ideas about how to potentially improve this, or do you see it not being an issue, even on large networks?

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+
+        storageInstances = new HashMap<>();
+        storagePositions = new HashSet<>();
+
+        Queue<BlockPos> searchQueue = new LinkedList<>();
+        Set<BlockPos> checked = new HashSet<>();
+
+        // check the posses around the controller
+        for (EnumFacing facing : EnumFacing.VALUES) {
+            searchQueue.add(getPos().offset(facing));
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:225]: To go off your todo comment about using the neighbor cache, this could first check `getNeighbor(facing)` is an instance of `IQuantumStorage`, and then only add the block pos to the searchQueue if that is true. However, I am unsure if the neighbor cache will be populated at this point, since this is called right after placement. That needs some checking.

**Review by ALongStringOfNumbers — COMMENTED** [review]
Just some code review. Want to actually check it in game as well, but will have to do so later.

Also I think it could be nice if the TOP stuff that I talked about in discord was implemented, but it doesn't have to be.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    @Override
+    public BlockPos getControllerPos() {
+        return controllerPos;
+    }
+
+    @Override
+    public void onRemoval() {
+        if (!getWorld().isRemote && isConnected()) {
+            IQuantumController controller = getQuantumController();
+            if (controller != null) controller.rebuildNetwork();
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:124]: the null check only exists because intellij warns about it

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+        super.onPlacement(placer);
+        if (getWorld() == null || getWorld().isRemote)
+            return;
+
+        // add to the network if an adjacent block is part of a network
+        // use whatever we find first, merging networks is not supported
+        tryFindNetwork();
+    }
+
+    private void tryFindNetwork() {
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:139]: yeah, the logic was copied to energy hatches. im thinking of consolidating this logic somehow.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+            this.controllerPos = null;
+            this.controller.clear();
+            if (this.connectedIcon != null) {
+                this.connectedIcon.setButtonTexture(GuiTextures.GREGTECH_LOGO_DARK);
+                this.connectedIcon.setTooltipText(null);
+            }
+            scheduleRenderUpdate();
+        } else if (dataId == GregtechDataCodes.LOCATE_CONTROLLER && buf.readBoolean()) {
+            // tell controller to highlight
+            BlockPosHighlightRenderer.renderBlockBoxHighLight(getControllerPos(), 6000, 1500);
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:211]: i don't see anything checking dimension in the renderer, so it might have the same issue

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityEnergyHatch.java [hunk 14]
```diff
@@ -204,4 +223,168 @@ public void doExplosion(float explosionPower) {
...
+        tryFindNetwork();
+        markDirty();
+    }
+
+    @Override
+    public void receiveCustomData(int dataId, PacketBuffer buf) {
+        super.receiveCustomData(dataId, buf);
+        if (dataId == GregtechDataCodes.UPDATE_CONTROLLER_POS) {
+            this.controllerPos = buf.readBlockPos();
+            this.controller.clear();
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityEnergyHatch.java:288]: i think the idea is that `controller` is lazily initialized. but i could do `new WeakReference<>(null)` instead. tbh i don't remember why i'm clearing it here in the first place.

[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityEnergyHatch.java:288]: actually, i'm clearing so the next time `getQuantumController()` is called, the reference is properly updated

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityCreativeChest.java [hunk 15]
```diff
@@ -193,4 +186,46 @@ public void receiveInitialSyncData(PacketBuffer buf) {
...
+        }
+
+        @NotNull
+        @Override
+        public ItemStack extractItem(int slot, int amount, boolean simulate) {
+            if (!active) return ItemStack.EMPTY;
+
+            ItemStack stack = super.getStackInSlot(slot);
+            stack.setCount(itemsPerCycle);
+            return simulate ? stack : stack.copy();
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityCreativeChest.java:214]: i only give a copy when not simulating so that mutating the stack doesn't affect the internal stack

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityCreativeChest.java [hunk 16]
```diff
@@ -72,7 +60,10 @@ public void renderMetaTileEntity(CCRenderState renderState, Matrix4 translation,
                 this);
         Textures.CREATIVE_CONTAINER_OVERLAY.renderSided(EnumFacing.UP, renderState, translation, pipeline);
         Textures.PIPE_OUT_OVERLAY.renderSided(this.getOutputFacing(), renderState, translation, pipeline);
-        Textures.ITEM_OUTPUT_OVERLAY.renderSided(this.getOutputFacing(), renderState, translation, pipeline);
+        if (!isConnected() && active) {
+            Textures.ITEM_OUTPUT_OVERLAY.renderSided(this.getOutputFacing(), renderState, translation, pipeline);
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityCreativeChest.java:64]: this is because a render update is not scheduled from clicking the active button

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java (hunk 3 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:239]: there would need to be a packet for when the quantum part is disconnected, to update the state of the icon

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+            if (mte instanceof IQuantumController quantumController) {
+                controller = new WeakReference<>(quantumController);
+                return quantumController;
+            } else {
+                // controller is no longer there for some reason, need to disconnect
+                setDisconnected();
+            }
+        }
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:106]: i'd need a reference to `CoverableView`, i'll see what i can do about it

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java (hunk 7 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:168]: the `rebuild` paramater is always false, idk what seni intended for this

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java (hunk 9 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:191]: there might be a case where a quantum part could connect to a new valid controller

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java (hunk 3 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:239]: nvm, i understand what you mean now

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java (hunk 9 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java]: actually moved the network checking out, so it's better now

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java (hunk 6 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:142]: this would be doable now that mui2 is merged into master (this pr was made before that)

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java:106]: consolidating this is problematic, since it needs a reference to a weak ref to the controller.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+                        GTValues.V[GTValues.HV] / 2 :
+                        GTValues.V[GTValues.LV] / 2;
+            }
+            case PROXY -> 8L;
+            case EXTENDER -> 2L;
+            case ENERGY -> 1L;
+        };
+    }
+
+    private void addEnergy(IQuantumStorage<?> storage) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:372]: This and the following methods are not used at all. Were they being used at one point?

File: src/main/java/gregtech/integration/theoneprobe/provider/QuantumEnergyProvider.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+        return ":quantum_energy_provider";
+    }
+
+    @Override
+    public void addProbeInfo(ProbeMode mode, IProbeInfo probeInfo, EntityPlayer player, World world,
+                             IBlockState blockState, IProbeHitData data) {
+        if (blockState.getBlock().hasTileEntity(blockState) &&
+                world.getTileEntity(data.getPos()) instanceof IGregTechTileEntity gtte) {
+            if (gtte.getMetaTileEntity() instanceof IQuantumController controller) {
+                configureEnergyUsage(controller.getEnergyUsage() / 10, probeInfo);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/integration/theoneprobe/provider/QuantumEnergyProvider.java:32]: Why is the energy usage here divided by 10, when the quantum controller just removes all of `getEnergyUsage` twice per second?

**Review by ALongStringOfNumbers — COMMENTED** [review]
Some information about the various status (Red, Blue, Not online) in TOP would be appreciated.

Additionally, adding a line or two to the scanner about energy consumpution and maybe number of connected item/fluid storages would be nice.

Had an issue with the quantum storage network not updating right away. I broke the energy hatch that was connected to the quantum network to test something, and the quantum network did not update until I placed an extender on a line of extenders, which forced a rescan, or when I attempted to export something from the network.

Items and fluids are still reported in AE2, when using a storage bus on the controller, but when the network is "offline" (All storages showing red and no power being provided to the network).

Similarly, it seems like the display status on the quantum components connected to the storage network is not updated until something tries to push into the network, or the network is refreshed via block update. From the case above, after I broke the energy hatch, I reconnected it, but the displays on the chests and tanks I had in my network did not update.

File: src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java (hunk 18 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorageController.java:372]: there were being used to add stuff to the controller like a list, but i forgot to remove them

File: src/main/java/gregtech/integration/theoneprobe/provider/QuantumEnergyProvider.java (hunk 19 above)
Comment (ghzdude) [reply src/main/java/gregtech/integration/theoneprobe/provider/QuantumEnergyProvider.java:32]: this displays the EU per tick, but the energy usage is per 10 ticks, so it's divided by 10

File: src/main/java/gregtech/api/capability/IQuantumStorage.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+
+    default boolean isConnected() {
+        // use controllerPos here because it is synced
+        // on both sides, where controller is not
+        return getControllerPos() != null;
+    }
+
+    default void tryFindNetwork() {
+        for (EnumFacing facing : EnumFacing.VALUES) {
+            if (getWorld().getBlockState(getPos().offset(facing)).getBlock() == Blocks.AIR) continue;
```
[the remark this answers is not in the document]
Comment (ghzdude) [reply src/main/java/gregtech/api/capability/IQuantumStorage.java:36]: are you talking about `Block#isAir()`? i don't see the method you mention. also `getNeighbor()` does not explicitly check for if the blockpos is air from what i can see

**Review by ALongStringOfNumbers — APPROVED** [review]
